findbugs/build.gradle

```diff
@@ -181,6 +181,40 @@ task annotationSourcesJar(type: Jar, dependsOn: classes) {
   include '**/annotations/*'
 }
 
+task testHarnessJar(type: Jar, dependsOn: testClasses) {
+  baseName 'test-harness'
+  from sourceSets.test.output
+  include 'edu/umd/cs/findbugs/AbstractIntegrationTest.class'
+  include 'org/sonar/plugins/findbugs/resource/SmapParser.class'
+  include 'edu/umd/cs/findbugs/test/**/*.class'
+}
+
+task testHarnessJavadoc(type: Javadoc) {
+  source sourceSets.test.allSource
+  include 'edu/umd/cs/findbugs/AbstractIntegrationTest.java'
+  include 'org/sonar/plugins/findbugs/resource/SmapParser.java'
+  include 'edu/umd/cs/findbugs/test/**/*.java'
+  destinationDir = file("${project.docsDir}/test-harness")
+  options.with {
+    tags = ['Test']
+  }
+}
+
+task testHarnessJavadocJar(type: Jar, dependsOn: testHarnessJavadoc) {
+  classifier = 'javadoc'
+  baseName 'test-harness'
+  from testHarnessJavadoc.destinationDir
+}
+
+task testHarnessSourcesJar(type: Jar, dependsOn: testClasses) {
+  classifier = 'sources'
+  baseName 'test-harness'
+  from sourceSets.test.allSource
+  include 'edu/umd/cs/findbugs/AbstractIntegrationTest.java'
+  include 'org/sonar/plugins/findbugs/resource/SmapParser.java'
+  include 'edu/umd/cs/findbugs/test/**/*.java'
```

Comment:
these 4 lines are repeated over and over... I'm particularly worried about the includes. We should refactor this by extracting all 3 ant matchers to an array in a variable and using that directly (`include` supports it). We may also want to start splitting this build file into one per artifact under `gradle` dir.

Something such as:

```
def harnessIncludes = ['edu/umd/cs/findbugs/AbstractIntegrationTest.class',
  'org/sonar/plugins/findbugs/resource/SmapParser.class',
  'edu/umd/cs/findbugs/test/**/*.class']

task testHarnessJar(type: Jar, dependsOn: testClasses) {
  baseName 'test-harness'
  from sourceSets.test.output
  include harnessIncludes
}

task testHarnessJavadoc(type: Javadoc) {
  source sourceSets.test.allSource
  include harnessIncludes
  destinationDir = file("${project.docsDir}/test-harness")
  options.with {
    tags = ['Test']
  }
}

task testHarnessJavadocJar(type: Jar, dependsOn: testHarnessJavadoc) {
  classifier = 'javadoc'
  baseName 'test-harness'
  from testHarnessJavadoc.destinationDir
}

task testHarnessSourcesJar(type: Jar, dependsOn: testClasses) {
  classifier = 'sources'
  baseName 'test-harness'
  from sourceSets.test.allSource
  include harnessIncludes
}

plugins.withType(org.gradle.api.publish.maven.plugins.MavenPublishPlugin) { // lazy as plugin may not applied yet
  publishing {
    publications {
      testHarness(MavenPublication) {
        artifactId 'test-harness'
        artifact testHarnessJar

        artifact testHarnessSourcesJar {
          classifier 'sources'
        }

        artifact testHarnessJavadocJar {
          classifier 'javadoc'
        }
      }
    }
  }
}
```

Also of note, this should probably be included in `artifacts { }` too, shouldn't it?